pkg/cloudproviders/k8s/k8s.go

```diff
@@ -0,0 +1,380 @@
+package k8s
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"os"
+	"regexp"
+	"strings"
+	"time"
+
+	"k8s.io/client-go/util/flowcontrol"
+
+	"github.com/ash2k/stager/wait"
+
+	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/atlassian/gostatsd"
+	"github.com/atlassian/gostatsd/pkg/util"
+	"github.com/sirupsen/logrus"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+
+	core_v1 "k8s.io/api/core/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	core_v1inf "k8s.io/client-go/informers/core/v1"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+)
+
+var (
+	// DefaultAnnotationTagWhitelist is the default ParamAnnotationTagWhitelist. Everything beginning with a standard
+	// gostatsd string is included by default.
+	DefaultAnnotationTagWhitelist = []string{
+		fmt.Sprintf("^%s", AnnotationPrefix),
+	}
+	// DefaultLabelTagWhitelist is the default ParamLabelTagWhitelist. Every label is ignored by default.
+	DefaultLabelTagWhitelist []string
+)
+
+const (
+	// ProviderName is the name of the k8s cloud provider.
+	ProviderName = "k8s"
+	// PodsByIPIndexName is the name of the index function storing pods by IP.
+	PodsByIPIndexName = "PodByIP"
+	// AnnotationPrefix is the annotation prefix that is turned into tags by default.
+	AnnotationPrefix = "gostatsd.atlassian.com/"
+	// NodeNameEnvVar is the name of the environment variable that the k8s cloud provider will find it's node name in.
+	NodeNameEnvVar = "GOSTATSD_NODE_NAME"
```

Comment:
All our configuration comes from viper. Why environment variables?

Reply:
It's a Kubernetes thing. I don't think we can easily do viper for this.

https://kubernetes.io/docs/tasks/inject-data-application/downward-api-volume-expose-pod-information/

I thought very hard about how I could put this into Viper before I committed to this. The only way I can think of would be to put an initialisation script into the container that would read the environment variable and insert it into the config.toml instead of a placeholder. This would put additional burden on the operator as the binary itself is no longer standalone for using inside k8s.

Reply:
Ok, I see. Let's keep it like this then. We can always change it.

Reply:
However I think it'd be better to not expose the function to get the value and only use it in the factory. This will prevent spreading the knowledge of where this value is coming from all over the place.

Reply:
Definitely agree with you about not spreading the knowledge beyond the factory.

Actually there is one possible way we could fix this. We could allow the config file to point to a file that contains the node name, and then create that file with a value from the downwards api. Both ways kind of depart from viper a bit though.

Do you think this would be a better fit @ash2k ?

Reply:
I think it's better to keep it as is.

Reply:
I disagree with both these approaches. If it's needed, it's on the service owner to provide it to viper. A wrapper script for config is trivial. By definition we aren't a standalone binary in a k8s environment, because there's all the k8s stuff to define what's even being ran.

Reply:
@tiedotguy that's a fair stance. We could use docs to show usage patterns, but I still disagree with using a wrapper script for config unless we're providing it inside a k8s-built container for the user.

How about we allow specifying the node name in config, which overrides environment variable?

Reply:
I would say make it required (log a descriptive error then fail, rather than `SetDefault`), and document it with a small snippet of what the config would look like. People should be reading documentation anyway, and the "fail fast" tells them something is missing even if they don't.

We can provide a sample script easily enough in the repo - I think it would be cool to include a full kubernetes yaml file as well to bring everything up, and this can just be in there with it.

I didn't have any definite plans to head in to the office, but I can swing past on Friday if you like (if you need this earlier I can make that happen too, Friday is just the typical day because it's a bit more relaxed)

Reply:
F2F would probably be the easiest, Friday is fine 🙂

Reply:
Can do!

Reply:
I was re-reading the thread with fresh eyes, trying to figure out what it is I didn't like about it. Essentially it was gostatsd changing its configuration model to conform, but I wasn't really considering it from the k8s API side.  The solution is ultimately a bridge between the two, which is what you suggested here:

> Actually there is one possible way we could fix this. We could allow the config file to point to a file that contains the node name, and then create that file with a value from the downwards api. Both ways kind of depart from viper a bit though.

Looking at it from that perspective, I've changed my opinion and think this is the right approach, however I would suggest using an env var defined by Viper (with a sane default), rather than a file, mainly because It's simpler code.

Thoughts?

Reply:
Yeah that sounds good to me. I'll try and implement it. To clarify, we're having a Viper defined env var that can also be overridden by config in the `.toml` right?

Reply:
Yeah, just like..

```
v.SetDefault("k8s-host-name-var", "GOSTATSD_BLAH")
...
hostName := GetEnv(v.GetString("k8s-host-name-var"))
```

... or something.

Reply:
I was thinking something more like
```
// replacer is to allow this to be a k8s sub option that conforms to env var naming standards
replacer := strings.NewReplacer(".", "_")
v.SetEnvKeyReplacer(replacer)
// this default could be sensible, or empty and fail-fast
v.SetDefault("k8s.node-name", "")
v.AutomaticEnv()

// this should retrieve from env var or config, depending on what's set and whatever the override order is
nodeName := v.Get("k8s.node-name")
```

Thoughts?

Reply:
The one downside is that particular variable in the config file would have to be named with `_` instead of `-` because we're using our replacer on the `.` to `_`

Reply:
Never mind, this probably won't work since there's a top level replacer already. I'll see what I can do to make something good within the constraints.

Reply:
Got something working here, just need to write tests for it now 🙂